vm/chef/cookbooks/postgresql/files/pgcluster-utils

```diff
@@ -0,0 +1,190 @@
+
+    echo "Trying to connect to instance ${instance}..."
+    while [[ "${connected}" == "false" ]]; do
+      # Avoid to fail script as we are checking exit-code.
+      set +e
+      su - postgres -c "psql -U bucardo -d ${database_name} -h \"${instance}\" -c \"SELECT 1 FROM ${sample_table};\"" \
+        >/dev/null 2>&1
+      set -e
+
+      if [[ "$?" -eq 0 ]]; then
+        echo "Connected to the instance: ${instance}."
+        connected="true"
+      else
+        echo "Connection to ${instance} failed."
+      fi
+    done
+  done
+}
+
+#######################################
+# Enroll all the instances databases to the replication service.
+# Arguments:
+#   Database name.
+#   Instances list.
+#######################################
+function add_databases_to_replication() {
+  local -r database_name="$1"
+  local -r instances="$2"
+  local instance_id=""
+  local node_name=""
+
+  for instance in $instances; do
+    # Node name should be short and without special chars.
+    instance_id="$(echo ${instance} | awk -F '-' '{ print $NF }')"
+    node_name="node${instance_id}"
+
+    echo "Adding ${instance} to bucardo..."
+
+    # Add database server
+    bucardo add database "${node_name}" \
+      dbname="${database_name}" \
+      host="${instance}"
+
+    # Add all available tables
+    bucardo add all table % db="${node_name}"
+  done
+}
+
+#######################################
+# Add the replication syncs permutating all possible combinations.
+# Arguments:
+#   Instances list.
+#######################################
+function add_replication_syncs() {
+  local -r instances_list="$1"
+  local source_instance_id=""
+  local source_node_name=""
+  local destination_node_name=""
+
+  for instance in $instances_list; do
+    source_instance_id="$(echo ${instance} | awk -F '-' '{ print $NF }')"
+    source_node_name="node${source_instance_id}"
+
+    echo "Adding ${instance} replicas:"
+    for replica in $instances_list; do
+      # Skip replication between same instance
+      if [[ "${replica}" == "${instance}" ]]; then
+        continue
+      fi
+
+      destination_instance_id="$(echo ${replica} | awk -F '-' '{ print $NF }')"
+      destination_node_name="node${destination_instance_id}"
+      herd_name="${source_node_name}_to_${destination_node_name}"
+      sync_name="sync_${herd_name}"
+
+      echo "Adding ${replica} as ${instance} replica. Herd: ${herd_name}. Sync: ${sync_name}"
+
+      # Add sync between source and destination
+      bucardo add all tables --herd="${herd_name}" --db="${source_node_name}"
+      bucardo add sync "${sync_name}" \
+        relgroup="${herd_name}" \
+        db="${source_node_name},${destination_node_name}"
+    done
+
+    echo "----"
+  done
+}
+
+#######################################
+# Configure the PostreSQL authentication for master.
+#######################################
+function configure_pg_hba_master() {
+  sed -i 's/peer/trust/g' "${pghba_path}"
+  configure_pg_hba_replica
```

Comment:
if `configure_pg_hba_replica` is applicable for all nodes, both master and replicas, maybe better go for 
`configure_pb_hba_node` for common stuff and master specific for this one:

```
function configure_pg_hba_master() {
  sed -i 's/peer/trust/g' "${pghba_path}"
  # WITHOUT this one: configure_pg_hba_replica
}
```

Then in `postgresql-cluster` can be like this:

```
echo "Add common pg_hba configuration..."
configure_pg_hba_node

if [[ "${current_node}" == "${master_hostname}" ]]; then
  # Configure master node
  echo "Setting up master node..."
  ...
  echo "Add pg_hba master configuration..."
  configure_pg_hba_master
else
    # Configure replica nodes
    ...
fi
```

Reply:
Hmmm part of the comment makes sense. But I can't use the common method out of the main if. As master node will await for all the replicas to be connected. And pghba config only will take effect after the restart. I'll implement the suggestion partially and resend for review :)

Reply:
Please, check new version